docs/docs/authentication-tutorial.md

```diff
@@ -2,7 +2,445 @@
+
+export const getUser = () =>
+  isBrowser() && window.localStorage.gatsbyUser
+    ? JSON.parse(window.localStorage.gatsbyUser)
+    : {}
+
+const setUser = user => (window.localStorage.gatsbyUser = JSON.stringify(user))
+
+export const handleLogin = ({ username, password }) => {
+  if (username === `john` && password === `pass`) {
+    return setUser({
+      username: `john`,
+      name: `Johnny`,
+      email: `johnny@example.org`,
+    })
+  }
+
+  return false
+}
+
+export const isLoggedIn = () => {
+  const user = getUser()
+
+  return !!user.username
+}
+
+export const logout = callback => {
+  setUser({})
+  callback()
+}
+```
+
+## Creating client-only routes
+
+At the beginning of this tutorial, you created a "default" Gatsby site, which includes the `@reach/router` library. Now, using the [@reach/router](https://reach.tech/router/) library, you can create routes available only to logged-in users. This library is used by Gatsby under the hood, so you don't even have to install it.
+
+First, edit `gatsby-node.js`. You will define that any route that starts with `/app/` is part of your restricted content and the page will be created on demand:
+
+```javascript:title=gatsby-config.js
+// Implement the Gatsby API “onCreatePage”. This is
+// called after every page is created.
+exports.onCreatePage = async ({ page, actions }) => {
+  const { createPage } = actions
+
+  // page.matchPath is a special key that's used for matching pages
+  // only on the client.
+  if (page.path.match(/^\/app/)) {
+    page.matchPath = "/app/*"
+
+    // Update the page.
+    createPage(page)
+  }
+}
+```
+
+> Note: There is a convenient plugin that already does this work for you: [gatsby-plugin-create-client-paths](https://www.gatsbyjs.org/packages/gatsby-plugin-create-client-paths/)
+
+Now, you must create a generic page that will have the task to generate the restricted content:
+
+```jsx:title=src/pages/app.js
+import React from "react"
+import { Router } from "@reach/router"
+import Layout from "../components/layout"
+import Profile from "../components/profile"
+import Login from "../components/login"
+
+const App = () => (
+  <Layout>
+    <Router>
+      <Profile path="/app/profile" />
+      <Login path="/app/login" />
+    </Router>
+  </Layout>
+)
+
+export default App
+```
+
+Next, add the components regarding those new routes:
+
+```jsx:title=src/components/profile.js
+import React from "react"
+
+const Profile = () => (
+  <>
+    <h1>Your profile</h1>
+    <ul>
+      <li>Name: Your name will appear here</li>
+      <li>E-mail: And here goes the mail</li>
+    </ul>
+  </>
+)
+
+export default Profile
+```
+
+```jsx:title=src/components/login.js
+class Login extends React.Component {
+  state = {
+    username: ``,
+    password: ``,
+  }
+
+  handleUpdate = event => {
+    this.setState({
+      [event.target.name]: event.target.value,
+    })
+  }
+
+  handleSubmit = event => {
+    event.preventDefault()
+    handleLogin(this.state)
```

Comment:
handleLogin needs to be imported, right?

Reply:
Agreed. As well as the whole import block that I skipped :disappointed_relieved: